src/Common/src/System/Security/Cryptography/Asn1V2.Serializer.cs

```diff
@@ -70,6 +74,52 @@ private static Deserializer TryOrFail<T>(TryDeserializer<T> tryDeserializer)
             };
         }
 
+        private static FieldInfo[] GetOrderedFields(Type typeT)
+        {
+            return s_orderedFields.GetOrAdd(
+                typeT,
+                t =>
+                {
+                    // https://github.com/dotnet/corefx/issues/14606 asserts that ordering by the metadata
+                    // token on a SequentialLayout will produce the fields in their layout order.
+                    //
+                    // Some other alternatives:
+                    // * Add an attribute for controlling the field read order.
+                    //    fieldInfos.Select(fi => (fi, fi.GetCustomAttribute<AsnFieldOrderAttribute>(false)).
+                    //      Where(val => val.Item2 != null).OrderBy(val => val.Item2.OrderWeight).Select(val => val.Item1);
+                    //
+                    // * Use Marshal.OffsetOf as a sort key
+                    //
+                    // * Some sort of interface to return the fields in a declared order, using either
+                    //   an existing object, or Activator.CreateInstance.  It would need to check that
+                    //   any returned fields actually were declared on the type that was queried.
+                    //
+                    // * Invent more alternatives
+                    FieldInfo[] fieldInfos = t.GetFields(FieldFlags);
+
+                    if (fieldInfos.Length == 0)
+                    {
+                        return Array.Empty<FieldInfo>();
+                    }
+
+                    try
+                    {
+                        int token = fieldInfos[0].MetadataToken;
+                    }
+                    catch (InvalidOperationException)
+                    {
+                        // If MetadataToken isn't available (like in ILC) then just hope that
+                        // the fields are returned in declared order.  For the most part that
+                        // will result in data misaligning to fields and deserialization failing,
+                        // thus a CryptographicException.
+                        return fieldInfos;
+                    }
+
+                    return fieldInfos.OrderBy(fi => fi.MetadataToken).ToArray();
```

Comment:
Could we just sort the original array instead, e.g.
```C#
Array.Sort(fieldInfos, (x, y) => x.MetadataToken.CompareTo(y.MetadataToken));
return fieldInfos;
```
?

Reply:
That seems way too sensible :smile:.